### DISCUSSION TITLE
No deserialization found for type dynamic

### DISCUSSION THREAD

**Discussion by onurcagilci** [serverpod/serverpod#2722] [Q&A] [answered]
I want to return an endpoint with Map<String, dynamic> and I am sending the model I created on the serverpod side as you can see below, but I am getting the error "No deserialization found for type dynamic". Can't I return Map<String, dynamic> as the return type? Or is there a problem while sending the model?

**Here my endpoint:**

 ```
Future<Map<String, dynamic>> getUserInfoByUserInfoId(
    Session session,
    int userInfoId,
    Map<String, dynamic>? deviceInfo,
  ) async {
    Map<String, dynamic> result = {};

    try {
      final userInfo = await EAllergyUser.db
          .findFirstRow(session, where: (u) => u.userId.equals(userInfoId));
      if (userInfo == null) {
        result = {
          "userInfo": null,
          "verificationCode": null,
          "isDetectedNewDevice": false,
          "message": "Kullanıcı bulunamadı",
          "statusCode": 404
        };
        return result;
      }

      if (userInfo.macAddress != null && userInfo.macAddress!.isEmpty && deviceInfo != null) {
        userInfo.macAddress!.add(deviceInfo['id']);
        await EAllergyUser.db.updateRow(session, userInfo);

        result = {
          "userInfo": userInfo,
          "verificationCode": null,
          "isDetectedNewDevice": false,
          "message": "Kullanıcı bilgileri başarıyla getirildi",
          "statusCode": 200
        };

        return result;
      } else {
        if (deviceInfo != null && !userInfo.macAddress!.contains(deviceInfo['id'])) {
          final mainUser = await UserInfo.db.findById(session, userInfo.userId);

          // Generate verification code
          String verificationCode = _generateVerificationCode();

          // Send email with deviceInfo and verification code
          await _sendVerificationEmail(session, mainUser!.email ?? "",
              deviceInfo, verificationCode, mainUser);

          // Store the verification code (implementation depends on your storage solution)
          await _storeVerificationCode(session, userInfo.userId, verificationCode);

          result = {
            "userInfo": userInfo,
            "verificationCode": verificationCode,
            "isDetectedNewDevice": true,
            "message": "Yeni bir cihaz girişi tespit edildi. Lütfen siz olduğunuzu doğrulayınız.",
            "statusCode": 200
          };

          return result;
        } else {
          result = {
            "userInfo": userInfo,
            "verificationCode": null,
            "isDetectedNewDevice": false,
            "message": "Kullanıcı bilgileri başarıyla getirildi",
            "statusCode": 200
          };

          return result;
        }
      }
    } catch (e) {
      log(e.toString(),
          time: DateTime.now(), name: "getUserInfoByUserInfoId", error: e);
      return {
        "userInfo": null,
        "verificationCode": null,
        "isDetectedNewDevice": false,
        "message": "Bir hata ile karşılaşıldı",
        "statusCode": 500
      };
    }
  }

```
**and console log**

No deserialization found for type dynamic
#0      SerializationManager.deserialize (package:serverpod_serialization/src/serialization.dart:124:5)
#1      Protocol.deserialize (package:eallergysl_server/src/generated/protocol.dart:1631:18)
#2      Protocol.deserialize.<anonymous closure> (package:eallergysl_server/src/generated/protocol.dart:1576:48)
#3      MapBase.map (dart:collection/maps.dart:82:28)
#4      Protocol.deserialize (package:eallergysl_server/src/generated/protocol.dart:1575:27)
#5      EndpointDispatch._formatArg (package:serverpod/src/server/endpoint_dispatch.dart:183:33)
#6      EndpointDispatch.handleUriCall (package:serverpod/src/server/endpoint_dispatch.dart:122:25)
<asynchronous suspension>
#7      Server._handleRequest (package:serverpod/src/server/server.dart:298:18)
<asynchronous suspension>

 -  _CLI Version: 2.0.2
 - _Library version: 2.0.2

IDE: Android Studio,
OS: MacOS Sonoma 14.6.1

Thanks.

**Comment by Isakdl** [comment, marked as answer]
`dynamic` is not supported as a type in either the models or endpoint params/return values. 

Why do you want to return this JSON manually? You could just as easily create a model like so:

```yaml
class: MyResponse
fields:
  userInfo: model:auth:userInfo?
  verificationCode: String?
  isDetectedNewDevice: bool
  message: String
  statusCode: int
```

If you use the generated class from the above definition, Serverpod will manage the Serialization for you. Hence you will not need to do any manual conversion and you get a typed interface!

See: https://docs.serverpod.dev/concepts/models

Endpoints also supports managing exceptions for you, you can simply throw a serializable exception and Serverpod will automatically send it to the client and throw it there! With this you could remove the statusCode field and simply throw an error instead in the error case.

E.g.

```yaml
exception: MyException
fields:
  message: String
```
See: https://docs.serverpod.dev/concepts/exceptions

A side note, it looks like you are building an endpoint for a specific user? If so I would recommend not sending the userId as a param. It means that anyone could send a request with someone else's userId and retrieve information about someone else, e.g. this is a security risk.

Instead, you can get the userId from the authenticated user by looking at the session object. 

Like so:

```dart
final auth = await session.authenticated;
final userInfoId = auth?.userId;
```
